plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+
+from django.core.files import File
+from django.core.files.storage import default_storage
+from django.db.models import Q
+
+from pulpcore.plugin.models import Artifact, ProgressBar
+
+
+async def query_existing_artifacts(in_q, out_q):
```

Comment:
This function contains multiple code blocks each doing something complex.  Recommend decomposing into smaller functions each doing *one* thing.  Better functional decomposition will be more readable, testable and maintainable.

Reply:
I overall agree, but I don't want to do big code rearranging now for a few reasons. I do agree though that the internals of the stages could be improved.

1) the code has undergone a lot of hand testing. If the stages get a decomposing refactor then it all has to be re hand tested. That is a lot of extra effort that I think we would do better to spend in other areas of Pulp3.

2) We should do refactoring after we have tests to ensure we don't regress and we don't spend a lot of time hand testing code over and over. I want to help write those tests even, but I can't write them right now.

3) Other PRs with significant diffs on top of this PR are already underway, especially for https://pulp.plan.io/issues/3812 so if this code starts changing a lot there will be a lot of conflicts to resolve.

Can we make these improvements (that I agree with) later?

Reply:
Yes.